networkit/cpp/embedding/LearnEmbedding.cpp

```diff
@@ -0,0 +1,219 @@
+ *
+ *  adapted and reimplemented from node2vec
+ *  part of snap [https://github.com/snap-stanford/snap]
+ *  Copyright (c) 2007-2019, Jure Leskovec (under BSD license)
+ *
+ *  see [https://arxiv.org/pdf/1607.00653v1.pdf]
+ *
+ */
+
+// networkit-format
+
+#include <algorithm>
+#include <fstream>
+#include <unordered_map>
+#include <vector>
+
+#include <networkit/auxiliary/Random.hpp>
+
+#include "AliasSampler.hpp"
+#include "BiasedRandomWalk.hpp"
+#include "LearnEmbedding.hpp"
+
+// Code from https://github.com/nicholas-leonard/word2vec/blob/master/word2vec.c
+// Customized for SNAP and node2vec
+
+namespace NetworKit {
+namespace Embedding {
+
+// Number of negative samples. Value taken from original word2vec code.
+constexpr int negSamN = 5;
+
+constexpr int maxExp = 6;
+
+// Learning rate for SGD. Value taken from original word2vec code.
+constexpr double startAlpha = 0.025;
+constexpr double minAlpha = startAlpha * 0.0001;
+
+using Vocab = std::vector<count>;
+
+Vocab learnVocab(AllWalks &allWalks, count nn) {
+    Vocab vocab(nn);
+    for (auto &walk : allWalks) { // for every walk starting from this node
+        for (auto &node : walk) { // for every node in this walk
+            ++vocab[node];
+        }
+    }
+    return vocab;
+}
+
+// Precompute unigram table using alias sampling method
+AliasSampler vocabSampler(Vocab &vocab) {
+    double trainWordsPow = 0;
+    auto sz = vocab.size();
+    std::vector<float> probV(sz); // NOT ... {sz}
+    for (count i = 0; i < sz; ++i) {
+        double e = std::pow(vocab[i], 0.75);
+        probV[i] = e;
+        trainWordsPow += e;
+    }
+    std::transform(probV.begin(), probV.end(), probV.begin(),
+                   [&trainWordsPow](float f) { return f / trainWordsPow; });
+
+    AliasSampler as(sz);
+    as.unigram(probV);
+    return as;
+}
+
+// Initialize negative embeddings
+Embeddings initNegEmb(Vocab &vocab, int dimensions) {
+    return Embeddings{vocab.size(), std::vector<float>(dimensions)};
+}
+
+// Initialize positive embeddings
+Embeddings initPosEmb(Vocab &vocab, int dimensions) {
+    Embeddings emb{vocab.size(), std::vector<float>(dimensions)};
+
+    for (index i = 0; i < emb.size(); ++i) {
+        std::generate(emb[i].begin(), emb[i].end(),
+                      [&]() { return (uniform_real() - 0.5) / dimensions; });
+    }
+    return emb;
+}
+
+void trainModel(ModelData &model, count walkNr) {
+    AllWalks &allWalks = model.allWalks;
+    count &wordCntAll = model.wordCntAll;
+    AliasSampler &as = model.as;
+    count allWords = model.allWords;
+    count iterations = model.iterations;
+    count dimensions = model.dimensions;
+    count winSize = model.winSize;
+    double alpha = model.alpha;
+    Embeddings &synPos = model.synPos;
+    Embeddings &synNeg = model.synNeg;
+
+    Walk &thisWalk = allWalks[walkNr];
+
+    for (index wordI = 0; wordI < thisWalk.size(); ++wordI) {
+        if (wordCntAll % 10000 == 0) {
```

Comment:
I am not very familiar with the algorithm but this 1000 seems a bit strange. Why this number? Can we store it into a global variable and document it?